File: Sources/SwiftyThemeObject.m
//
//  SwiftyThemeObject.m
//  SwiftyTheme
//
//  Created by apple on 2020/5/6.
//  Copyright © 2020 yinhe. All rights reserved.
//

#import "SwiftyThemeObject.h"

@interface SwiftyThemeObject()
@property (nonatomic, copy) NSString *sel;
@property (nonatomic, strong) NSArray *args;
@end

@implementation SwiftyThemeObject

- (instancetype)initWithSel:(NSString *)sel args:(NSArray *)args
{
    self = [super init];
    if (self) {
        self.sel = sel;
        self.args = args;
    }
    return self;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        
    }
    return self;
}

- (NSUInteger)hash{
    long v1 = (long)sel_getName(NSSelectorFromString(self.sel));
    long v2 = (long)self.args;
    
    NSUInteger value = v1 ^ v2;
    
    return value;
}

- (BOOL)isEqual:(id)object{
    if (![object isKindOfClass:[self class]]) {
        return NO;
    }
    SwiftyThemeObject *other = object;
    
    if (other == self) {
        return YES;
    }
    
    BOOL r1 = self.sel == other.sel;
    
    BOOL r2 = YES;
    if (self.args.count == other.args.count) {
        for (int i = 0; i < self.args.count; i ++) {
            if (self.args[i] != other.args[i]) {
                r2 = NO;
                break;
            }
        }
    } else {
        r2 = NO;
    }
    
    return r1 && r2;
}
@end

File: Sources/SwiftyThemeObject.swift
//
//  SwiftyThemeObject.swift
//  SwiftyTheme
//
//  Created by apple on 2020/5/7.
//  Copyright © 2020 yinhe. All rights reserved.
//

import Foundation
import UIKit


internal struct SwiftyThemeColorKey {
    let key: String?
    init(key: String?) {
        self.key = key
    }
}

internal struct SwiftyThemeImageKey {
    let key: String?
    init(key: String?) {
        self.key = key
    }
}

public struct SwiftyThemeObject {
    let isEmpty: Bool
    var selector: String
    var args: [AnyObject?]
    
    init(selector: String, args: [AnyObject?], isEmpty: Bool) {
        self.isEmpty = isEmpty
        self.selector = selector
        self.args = args
    }
}

extension SwiftyThemeObject: Equatable {
    public static func == (lhs: SwiftyThemeObject, rhs: SwiftyThemeObject) -> Bool {
        
        let res1 = lhs.selector == rhs.selector

        var res2: Bool = true
        if lhs.args.count == rhs.args.count {
            for (index, arg) in lhs.args.enumerated() {
                if let arg1 = arg,
                    let arg2 = rhs.args[index] {
                    if !(arg1.isEqual(arg2)) {
                        res2 = false
                        break
                    }
                } else {
                    res2 = false
                    break
                }
            }
        } else {
            res2 = false
        }

        return res1 && res2
        
        //return lhs.selector == rhs.selector
    }
}
